draft-irtf-cfrg-opaque.md

```diff
@@ -1639,87 +1639,67 @@ applications can use to control OPAQUE:
-In this case, servers can only generate a single protocol error.
-
-| Name                     | Stage        | Reason                  |
-|:-------------------------|:-------------|:------------------------|
-| InvalidClientAkeMACError | ServerFinish | Invalid client AKE MAC. |
-
-### Application errors
-
-Implementations MUST verify certain input arguments and return an appropriate error. Note that
-these are not protocol errors, and are indicative. The following table enumerates server errors that can occur, where they occur, and the reason for the error.
+Finally, note that online guessing attacks (against any aPAKE) can be done from
+both the client side and the server side. In particular, a malicious server can
+attempt to simulate honest responses in order to learn the client's password.
+Implementations and deployments of OPAQUE SHOULD consider additional checks to
+mitigate this type of attack: for instance, by ensuring that there is a
+server-authenticated channel over which OPAQUE registration and login is run.
 
-| Name                        | Role   | Stage      | Reason                                         |
-|:----------------------------|:-------|:-----------|:-----------------------------------------------|
-| InvalidServerSecretKeyError | Server | ServerInit | Invalid server secret key.                     |
-| InvalidServerPublicKeyError | Server | ServerInit | Invalid server public key.                     |
-| InvalidOPRFSeedLengthError  | Server | ServerInit | Invalid OPRF seed length.                      |
-| InvalidEnvelopeLengthError  | Server | ServerInit | Invalid envelope length for the configuration. |
+## Error Considerations
+
+Some functions included in this specification are fallible. For example, the
+authenticated key exchange protocol may fail because client's password was
+incorrect or the authentication check failed, yielding an error. The explicit
+errors generated throughout this specifiation, along with conditions that lead
+to each error, are as follows:
+
+- EnvelopeRecoveryError: The envelope Recover function failed to produce any
+  authentication key material; {{envelope-recovery}}.
+- ServerAuthenticationError: The client failed to complete the authenticated
+  key exchange protocol with the server; {{ake-client}}.
+- ClientAuthenticationError: The server failed to complete the authenticated
+  key exchange protocol with the client; {{ake-server}}.
+
+Beyond these explicit errors, OPAQUE implementations can produce implicit errors.
+For example, if protocol messages sent between client and server do not match
+their expected size, an implementaton should produce an error. More generally,
+if any protocol message received from the peer is invalid, perhaps because the
+message contains an invalid public key (indicated by the AKE DeserializeElement
+function failing) or an invalid OPRF element (indicated by the OPRF DeserializeElement),
+then an implementation should produce an error.
```

Comment:
I think we should also mention implicit errors originating from VOPRF (e.g. InvalidInputError) and hash-to-curve (if any)

Reply:
`DeserializeElement` is one of those implicit OPRF errors. I don't think we need to be exhaustive with the list of implicit things here.

I plan to address InvalidInputError and how OPAQUE should deal with it in response to #312.